src/core/shape_inference/include/space_to_batch_shape_inference.hpp

```diff
@@ -15,75 +15,72 @@ namespace ov {
 namespace op {
 namespace v1 {
 
-template <class T>
-void shape_infer(const ov::op::v1::SpaceToBatch* op,
-                 const std::vector<T>& input_shapes,
-                 std::vector<T>& output_shapes,
-                 const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {}) {
-    using ValType = typename std::iterator_traits<typename T::iterator>::value_type::value_type;
-    NODE_VALIDATION_CHECK(op, input_shapes.size() == 4 && output_shapes.size() == 1);
+template <class TShape>
+std::vector<TShape> shape_infer(const SpaceToBatch* op,
+                                const std::vector<TShape>& input_shapes,
+                                const std::map<size_t, HostTensorPtr>& constant_data = {}) {
+    using TVal = typename TShape::value_type::value_type;
+    NODE_VALIDATION_CHECK(op, input_shapes.size() == 4);
 
     const auto& data_shape = input_shapes[0];
     const auto& block_shape = input_shapes[1];
     const auto& pads_begin_shape = input_shapes[2];
     const auto& pads_end_shape = input_shapes[3];
-    const ov::Rank data_rank = data_shape.rank();
-    bool got_const_data = false;
 
     auto inputs_same_ps = pads_begin_shape;
-    NODE_VALIDATION_CHECK(op,
-                          T::merge_into(inputs_same_ps, pads_end_shape) && T::merge_into(inputs_same_ps, block_shape),
-                          "block_shape, pads_begin and pads_end inputs must have the same shape. Got: ",
-                          block_shape,
-                          ", ",
-                          pads_begin_shape,
-                          " and ",
-                          pads_end_shape);
+    NODE_VALIDATION_CHECK(
+        op,
+        TShape::merge_into(inputs_same_ps, pads_end_shape) && TShape::merge_into(inputs_same_ps, block_shape),
+        "block_shape, pads_begin and pads_end inputs must have the same shape. Got: ",
+        block_shape,
+        ", ",
+        pads_begin_shape,
+        " and ",
+        pads_end_shape);
 
     NODE_VALIDATION_CHECK(op,
                           inputs_same_ps.rank().compatible(1),
                           "block_shape and pads inputs must have rank 1. Got: ",
                           inputs_same_ps.rank());
 
-    if (data_rank.is_static()) {
+    if (data_shape.rank().is_static()) {
```

Comment:
Does it work with `ov::Shape`?

Reply:
The shape_infer is used tested with StaticShape and PartialShape. Right no these shape_infer functions are not used with ov::Shape.
I think will not work for ov::Shape as this class interface is not compatible with StaticShape and PartialShape it will required some specialization or make this class compatible. There is more member functions which not exist in ov::Shape.